complete fix for halloween problem

yingjunwu opened the pull request:
When performing an update operation that looks up tuples via non-primary-key columns and modifies these tuple's primary key columns, a bug may occur (Details can be found here: https://github.com/cmu-db/peloton/issues/106).

To fix this bug, I added a statement-level write set to track any newly inserted tuples (rather than versions) created by the update executor.

@pmenon I modified the codegen code. Please check whether it makes sense.
@poojanilangekar @pervazea you may check whether my PR fix the previously reported bug.

poojanilangekar commented:
@yingjunwu Yes, it fixes the bug! 

But there's still an issue in the storage layer which will end up leaking memory eventually. The problem is that line 388 of data_table.cpp allocates a new tuple [using `GetEmptyTupleSlot`] and then the Insert fails because it violates the index constraint. So InsertTuple retuned `INVALID_ITEMPOINTER` [line 346 of data_table.cpp], creating an orphan tuple. I think this tuple should also be added to the transaction context's ReadWriteSet so that it needs to be recycled. 

Currently when I turned on LOG_TRACE, I noticed that the `ResetTuple` function in transaction_level_gc_manager.cpp is never called on such tuples. (It is called on the deleted versions as expected, so we know it is recycling tuples which were added to the `ReadWriteSet`). This probably explains why @pervazea and I found tuples from aborted transactions in the `ReadWriteSet`.

pervazea commented:
First, as Pooja states, this fixes the problem seen via the psql test. Thanks!

Second, I was thinking this was also going to fix issue #1223 but it does not. Should it? Some additional cases need to be covered by the write set?

yingjunwu commented:
@poojanilangekar this PR is not intended to fix the memory leak problem. But I can take a look at it.

@pervazea Well, I confronted the same problem when debugging the system. I thought it should be fixed in other PRs, and this PR was not going to address this problem. But I can take a look at it.

yingjunwu commented:
@pmenon okay will do.

I am not sure how you will parallelize queries in codegen. Maybe we could have a discussion if you confront any problems there.

yingjunwu commented:
@pmenon I've added comments to explain why statement-level write set is necessary. This PR should be ready to go.
@poojanilangekar @pervazea I will take care of the memory leak and the column disordering problem in separate PRs.

yingjunwu commented:
@poojanilangekar According to @apavlo, we should use concurrent set in order to support parallel query execution in the future. I think it is okay to leave it there and we may change it back to `std::set` later. Sure I will add a test case for the bug.

poojanilangekar commented:
@yingjunwu Can you use the `tbb::concurrent_unordered_multiset`. The reason I am suggesting this is that it would avoid people from using this as a `ReadWriteSet` sometime in the future.

yingjunwu commented:
@poojanilangekar I've changed to tbb set. It's not easy to create a test case for the bug, as it totally depends on which slot in the tile group the transaction is going to use to hold new versions/tuples. I will think about it later. You could merge this PR first.

yingjunwu commented:
@poojanilangekar @pervazea I updated `basic1.sql` and `psql_test.sh` to force testing halloween problem in the testing script. Instead of setting `active_tilegroup_count` to a magic number (as we did before), we now allow the user to explicitly set this number (default value is 1). This gives user better control over the space usage and parallelism. The statements I added to `basic1.sql` will trigger assertion once halloween problem happens. I think my test case should satisfy the requirement.

pervazea commented:
From email side discussion:
_We will not confront Halloween problem when the update field(s) are not primary keys. This is because we lookup targeted tuple versions through version chains. If the updated fields are not primary keys, then the newly created version will be appended to the end of the version chain, and the current transaction will not read this version._

Tests run on my laptop.

At:
commit 472d70636fcafccd1c16ebc226c6164c0e6b9f2f
Author: Yingjun Wu <yingjun.wu@ibm.com>
Date:   Wed Mar 28 10:57:56 2018 -0700

    fix bug in seq scan

via psql
```
default_database=# create table foo(id integer, year integer);
CREATE TABLE
default_database=# insert into foo values(10,1000);
INSERT 0 1
default_database=# select * from foo;
 id | year 
----+------
 10 | 1000
(1 row)

default_database=# update foo set id=id/2;
UPDATE 2
default_database=# select * from foo;
 id | year 
----+------
  2 | 1000
(1 row)
```
This is the Halloween problem, correct? No primary key. 

Now with latest version that includes tbb::concurrent_unordered_multiset
commit 0c5d41b9592aec0845559ffe5333a03472d1d846
Author: Yingjun Wu <yingjun.wu@ibm.com>
Date:   Tue Apr 3 22:23:09 2018 -0700

    update psql_ref_out.txt

again via psql

```
default_database=# create table foo(id integer, year integer);
CREATE TABLE
default_database=# insert into foo values(10,1000);
INSERT 0 1
default_database=# update foo set id=id/2;
UPDATE 1
default_database=# select * from foo;
 id | year 
----+------
  5 | 1000
(1 row)
```
Much better. No Halloween problem.

Couple of questions:
1. We were hitting symptoms that look like the Halloween problem, on a field that is not the primary key. Is this something different?

2. How does tbb::concurrent_unordered_multiset fix the issue observed? Are we just getting lucky?

yingjunwu commented:
@pervazea I have already fixed the halloween problem (fully). Please take a look.Thanks!

yingjunwu commented on src/common/init.cpp:
```diff
@@ -48,9 +50,12 @@ void PelotonInit::Initialize() {
-  storage::DataTable::SetActiveTileGroupCount(parallelism);
-  storage::DataTable::SetActiveIndirectionArrayCount(parallelism);
+  // require user to set parallel tilegroup count
+  PARALLEL_TILEGROUP_COUNT = settings::SettingsManager::GetInt(
```
I think we could leave it there. The `parallel` parameter was discussed long time ago and it controls the system scalability. It does not determine the number of tile groups. Check our MVCC paper for more detail. Originally we set parameter to a magic number as we expect the `brain` could eventually determine the parameter itself.

yingjunwu commented on script/testing/dml/basic1.sql:
```diff
@@ -10,6 +10,9 @@ CREATE TABLE foo(id integer PRIMARY KEY, year integer);
 
 -- load in the data
 
+INSERT INTO foo VALUES(10, 800);
```
Maybe. But adding more test cases will not harm the system robustness. I don't think it's necessary to remove this statement.

tcm-marcel commented on src/planner/update_plan.cpp:
```diff
@@ -31,6 +31,7 @@ UpdatePlan::UpdatePlan(storage::DataTable *table,
       auto col_id = target.first;
       update_primary_key_ =
           target_table_->GetSchema()->GetColumn(col_id).IsPrimary();
+      std::cout << "col_id = " << col_id << " " << update_primary_key_ << std::endl;
```
@yingjunwu The source validator complains, that `std::cout` is not allowed. [1] Is this a debug line?

[1] https://travis-ci.org/cmu-db/peloton/jobs/362378845#L5757

yingjunwu replied:
@tcm-marcel yes nice catch! I've fixed it. Thanks!

yingjunwu commented:
@pervazea @poojanilangekar @tcm-marcel addressed the snapshot isolation and in-place update problem. Please take a look. Hope this time we can fully tackle the halloween problem!

pmenon commented:
@yingjunwu Can you describe the fixes you've added to this PR related to SI and in-place updates? We should also consider renaming the PR to reflect the changes.

yingjunwu commented:
@pmenon This PR provides a complete fix for halloween problem. @poojanilangekar's PR cannot fix the halloween problem under snapshot isolation. Snapshot isolation retrieves the snapshot version by jumping from the latest version of a tuple. @poojanilangekar's fix added the latest version (rather than the snapshot version) into the statement-level write set. Also, it unnecessarily added to-be-updated version into the statement-level write set when in-place update (which indicates a single tuple is updated in a transaction by multiple times) occurs. 

I think my fix has provided sufficient comments. Hope this explanation helps!

poojanilangekar commented:
@yingjunwu, the only patch needed to fix that is the [87238c2](https://github.com/cmu-db/peloton/pull/1240/commits/87238c22aeacf5c30cb9a99c62da6a0dd8ddfe3d). The other changes need to be reverted. It was discussed with Andy and we don't want to change the psql test and add the `parallel_tile_group_count` variable. So I would suggest you either revert all the other changes. Or the easier option is that you can cherrypick just that commit and raise a new PR. Thanks!

yingjunwu commented:
I think we could fix that bug in another PR.
